auth/auth_test.go

```diff
@@ -697,25 +698,88 @@ func TestVerifyIDTokenWithNoProjectID(t *testing.T) {
 	}
 }
 
-func TestVerifyIDTokenInEmulatorMode(t *testing.T) {
-	os.Setenv(emulatorHostEnvVar, "localhost:9099")
-	defer os.Unsetenv(emulatorHostEnvVar)
+func TestVerifyIDTokenUnsigned(t *testing.T) {
+	token := getEmulatedIDToken(nil)
+
+	client := &Client{
+		baseClient: &baseClient{
+			idTokenVerifier: testIDTokenVerifier,
+		},
+	}
+	_, err := client.VerifyIDToken(context.Background(), token)
+	if !IsIDTokenInvalid(err) {
+		t.Errorf("VerifyIDToken(Unsigned) = %v; want = IDTokenInvalid", err)
+	}
+}
+
+func TestEmulatorVerifyIDToken(t *testing.T) {
+	s := echoServer(testGetUserResponse, t)
+	defer s.Close()
+
+	s.Client.idTokenVerifier = testIDTokenVerifier
+	s.Client.isEmulator = true
+
+	token := getEmulatedIDToken(nil)
+	ft, err := s.Client.VerifyIDToken(context.Background(), token)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	now := testClock.Now().Unix()
+	if ft.AuthTime != now-100 {
+		t.Errorf("AuthTime = %d; want = %d", ft.AuthTime, now-100)
+	}
+	if ft.Firebase.SignInProvider != "custom" {
+		t.Errorf("SignInProvider = %q; want = %q", ft.Firebase.SignInProvider, "custom")
+	}
+	if ft.Firebase.Tenant != "" {
+		t.Errorf("Tenant = %q; want = %q", ft.Firebase.Tenant, "")
+	}
+	if ft.Claims["admin"] != true {
+		t.Errorf("Claims['admin'] = %v; want = true", ft.Claims["admin"])
+	}
+	if ft.UID != ft.Subject {
+		t.Errorf("UID = %q; Sub = %q; want UID = Sub", ft.UID, ft.Subject)
+	}
+}
+
+func TestEmulatorVerifyIDTokenExpiredError(t *testing.T) {
+	s := echoServer(testGetUserResponse, t)
+	defer s.Close()
+
+	s.Client.idTokenVerifier = testIDTokenVerifier
+	s.Client.isEmulator = true
+
+	now := testClock.Now().Unix()
+	token := getEmulatedIDToken(mockIDTokenPayload{
+		"iat": now - 1000,
+		"exp": now - clockSkewSeconds - 1,
+	})
+
+	_, err := s.Client.VerifyIDToken(context.Background(), token)
+	if !IsIDTokenExpired(err) {
+		t.Errorf("VerifyIDToken(Expired) = %v; want = IDTokenExpired", err)
+	}
+}
 
+func TestEmulatorVerifyIDTokenUnreachableEmulator(t *testing.T) {
 	conf := &internal.AuthConfig{
-		ProjectID: "",
 		Opts:      optsWithTokenSource,
+		ProjectID: testProjectID,
+		Version:   testVersion,
 	}
-	c, err := NewClient(context.Background(), conf)
+	client, err := NewClient(context.Background(), conf)
 	if err != nil {
 		t.Fatal(err)
 	}
+	client.httpClient.Client.Transport = eConnRefusedTransport{}
+	client.isEmulator = true
 
-	defer func() {
-		if r := recover(); r == nil {
-			t.Error("VeridyIDToken() didn't panic")
-		}
-	}()
-	c.VerifyIDToken(context.Background(), testIDToken)
+	token := getEmulatedIDToken(nil)
+	_, err = client.VerifyIDToken(context.Background(), token)
+	if err == nil || !strings.HasPrefix(err.Error(), "failed to establish a connection") {
```

Comment:
Can we also test for the error code in this scenario? What's the error code that gets set in this situation?